### ISSUE TITLE
FileCache sometimes fails with binary formats due to incorrect assumption that binary files cannot be read as utf-8 text

### ISSUE THREAD

**Issue by rasmuse** [reclosedev/requests-cache#507]
### tl;dr:
First of all, thanks for a useful library with very nice documentation.

Now to the problem I have just identified. In the source code of `FileDict` it is assumed in the following place that the proper reading mode (binary or text) for a cache file can be identified based on whether the file can be decoded as utf-8. This assumption is not true and therefore the FileCache sometimes fails.

https://github.com/reclosedev/requests-cache/blob/v0.9.1/requests_cache/backends/filesystem.py#L120

I have no idea what is the best way to fix this, but my suggestion would be to explicitly assign the `is_binary` based on which serializer is being used (i.e., explicitly say `is_binary` for `pickle` and `bson` etc).

### The problem
I am using requests-cache v0.9.1 with a FileCache. Using `pickle` or `bson` backends I sometimes get error messages when there is a cache hit, and an attempted deserialization fails. Here is an example:
```
ERROR:requests_cache.backends.base:Unable to deserialize response with key ed4e9f5d9c16e0e5: a bytes-like object is required, not 'str'
```
In these cases, the page is downloaded again and the cache overwritten. When doing this multiple times, the problem sometimes goes away and the FileCache happily delivers cached responses again.

### Expected behavior
The deserialization should always work and the page should not be downloaded again.

### Steps to reproduce the behavior
Here is a script that often, but not always, reproduces the behavior.

```python
from pathlib import Path
import requests
import requests_cache
from requests_cache import CachedSession

url = "https://www.example.com"
request = requests.Request("GET", url).prepare()
cache_key = requests_cache.cache_keys.create_key(request)


def can_file_be_read_as_text(path):
    try:
        with open(path, "r") as f:
            f.read()
        return True
    except UnicodeDecodeError:
        return False


def create_or_open_filecache_and_use_it(serializer_name):
    backend = requests_cache.filesystem.FileCache(
        f"{serializer_name}_test",
        serializer=serializer_name,
        extension=serializer_name,
    )
    session = CachedSession("test", backend=backend)
    file_path = session.cache.cache_dir / f"{cache_key}.{serializer_name}"
    if file_path.exists():
        print("Readable as text", can_file_be_read_as_text(file_path))
    else:
        print("No cache file to read")
    response = session.get(url)
    print("Response type", type(response))


for serializer_name in ["json", "bson", "pickle"]:
    print("Serializer", serializer_name)
    create_or_open_filecache_and_use_it(serializer_name)
    create_or_open_filecache_and_use_it(serializer_name)
    print()
```

Typical output:

```
Serializer json
No cache file to read
Response type <class 'requests.models.Response'>
Readable as text True
Response type <class 'requests_cache.models.response.CachedResponse'>

Serializer bson
No cache file to read
Response type <class 'requests.models.Response'>
Readable as text False
Response type <class 'requests_cache.models.response.CachedResponse'>

Serializer pickle
No cache file to read
Response type <class 'requests.models.Response'>
Readable as text True
Unable to deserialize response with key ed4e9f5d9c16e0e5: a bytes-like object is required, not 'str'
Response type <class 'requests.models.Response'>
```

Note how `pickle` fails and `bson` works here. It could be the other way around. Also note how the failure of pickle coincides with `Readable as text True`.

### Workarounds
Well, use `json`, I suppose, since the default assumption is that `is_binary==False`.

### Environment
- requests-cache version: `0.9.1`
- Python version: `3.9`
- Platform: Windows 10

**Comment by JWCook** [comment]
Thanks for the thorough bug report!

At one time I think the behavior was as you described, setting `is_binary` based on the serializer, and a later PR changed that to simplify the code a bit. When reviewing that it didn't occur to me that a binary file could coincidentally be a valid UTF-8 text file.

As another temporary workaround you could manually set `is_binary`:
```python
session = CachedSession(backend='filesystem', serializer='pickle')
session.cache.responses.is_binary = True
```

To fix this, there may be a better generic way to detect binary filetypes, like maybe the [mimetypes](https://docs.python.org/3/library/mimetypes.html) module. Otherwise we can revert back to hardcoding this per serializer.

**Comment by rasmuse** [comment]
Thanks! I'm using your suggested workaround for now.

Unless it creates unacceptable backwards-compatibility problems for custom serializer pipelines, my clear preference would be to 
set `is_binary` based on the serializer. Explicit, clear, and will never fail once configured correctly. Using an automatic detection like `mimetypes` would frankly in my eyes only be a small improvement, because there will always be edge cases, especially considering the possibility of custom serializer pipelines.

**Comment by JWCook** [comment]
This is now fixed in v0.9.2.

**Comment by rasmuse** [comment]
Thank you for this!

Out of curiosity I checked the code edits in the PR and I just noted that the preconfigured `bson` serializer is being created with `is_binary=False`, here: https://github.com/reclosedev/requests-cache/pull/529/commits/b5bac5aae0213b44fe02e362ece9d4ef60ff8793#diff-ebf882e0673edc1bf8f1fa0ca5e45d5454e1d6d3b58c9f3babec75865d3f6fd5R95 

I think it should be `is_binary=True` there, right?

**Comment by JWCook** [comment]
Aha, you're right, this was related to a mistake I made in handling differences between pymongo's `bson` module and the standalone `bson` codec. I was using pymongo's [bson.json_util.dumps/loads](https://pymongo.readthedocs.io/en/stable/api/bson/json_util.html) for compatibility with the non-pymongo [bson.dumps/loads](https://github.com/py-bson/bson). TL;DR: `json_util` handles bson-supported types but returns plaintext, not binary.

This is what happens when you don't use MongoDB for a year+ and come back to it...

**Comment by rasmuse** [comment]
Oh dear. Thanks for explaining that! I had no idea about the `json_util` behavior, which I guess made it easier for me to spot the mistake.

**Comment by JWCook** [comment]
Alright, that has been fixed in the latest pre-release build. Thanks for the sanity check!
